src/core/infer_stats.h

```diff
@@ -0,0 +1,183 @@
+  // Optional 'parent' can be provided if the trace object has a parent.
+  // Model name, model version, and trace manager should be set before calling
+  // this function. And each ModelInferStats instance should not call this
+  // function more than once.
+  void NewTrace(Trace* parent = nullptr);
+
+  // Get the trace object associated to the inference.
+  // Return nullptr if the inference will not be traced or if NewTrace()
+  // has not been called.
+  Trace* GetTrace() const { return trace_; }
+
+
+ private:
+
+  // The trace manager associated with these stats. This object is not owned by
+  // this ModelInferStats object and so is not destroyed by this object.
+  OpaqueTraceManager* trace_manager_;
+
+  // The trace associated with these stats. This object is not owned by
+  // this ModelInferStats object and so is not destroyed by this object.
+  Trace* trace_;
+};
+#endif
+
+// A stats aggregator used for one backend.
+class StatsAggregator {
+ public:
+  struct InferStats {
+    InferStats()
+        : failure_count_(0), failure_duration_ns_(0), success_count_(0),
+          request_duration_ns_(0), queue_duration_ns_(0),
+          compute_input_duration_ns_(0), compute_infer_duration_ns_(0),
+          compute_output_duration_ns_(0)
+    {
+    }
+    uint64_t failure_count_;
+    uint64_t failure_duration_ns_;
+
+    uint64_t success_count_;
+    uint64_t request_duration_ns_;
+    uint64_t queue_duration_ns_;
+    uint64_t compute_input_duration_ns_;
+    uint64_t compute_infer_duration_ns_;
+    uint64_t compute_output_duration_ns_;
+  };
+
+  struct InferBatchStats {
+    InferBatchStats()
+        : count_(0), compute_input_duration_ns_(0),
+          compute_infer_duration_ns_(0), compute_output_duration_ns_(0)
+    {
+    }
+    uint64_t count_;
+    uint64_t compute_input_duration_ns_;
+    uint64_t compute_infer_duration_ns_;
+    uint64_t compute_output_duration_ns_;
+  };
+
+  // Create an aggregator for model statistics
+  StatsAggregator() : last_inference_ms_(0) {}
+
+  // Create an aggregator with metric reporter attached for model statistics
+  StatsAggregator(const std::shared_ptr<MetricModelReporter>& metric_reporter)
+      : last_inference_ms_(0), metric_reporter_(metric_reporter)
+  {
+  }
+
+  uint64_t LastInferenceMs() const { return last_inference_ms_; }
+  const InferStats& ImmutableInferStats() const { return infer_stats_; }
+  const std::map<size_t, InferBatchStats>& ImmutableInferBatchStats() const
+  {
+    return batch_stats_;
+  }
+
+  // FIXME passing device is somewhat confusing here, that is for updating
+  // metrics properly, but is here the good place for updating matrics?
+  //
+  // Add durations to Infer stats for a failed inference request.
+  void UpdateFailedInferStats(
+      int device, uint64_t last_timestamp_ms, uint64_t request_duration_ns);
+
+  // Add durations to infer stats for a successful inference request.
+  void UpdateSuccessInferStats(
```

Comment:
Have a UpdateSuccess() function that takes all these args (minus device, plus batch_size) and takes the lock once to update both infer and batch stats.